pasqal_cloud/client.py

```diff
@@ -230,18 +206,22 @@ def get_batch_jobs(self, batch_id: str) -> list[Dict[str, Any]]:
             # the same order as they do in the GET /batches/{id} response
         )
 
+    @retry_http_error(max_retries=2)
+    def _download_results(self, results_link: str) -> JobResult:
+        response = requests.get(results_link)
+        response.raise_for_status()
+        data = response.json()
+        return JobResult(
+            raw=data.pop("raw", None), counter=data.pop("counter", None), **data
+        )
+
     def get_job_results(self, job_id: str) -> Optional[JobResult]:
         results_link = self._request(
             "GET", f"{self.endpoints.core}/api/v1/jobs/{job_id}/results_link"
         )["data"]["results_link"]
         if results_link:
             try:
-                response = requests.get(results_link)
-                response.raise_for_status()
-                data = response.json()
-                return JobResult(
-                    raw=data.pop("raw", None), counter=data.pop("counter", None), **data
-                )
+                return self._download_results(results_link)
             except HTTPError:
                 pass
```

Comment:
In case of an HTTPError even after the retry, do you think we should raise the exception or just set the results to None?

The reason I chose to return None is that we currently set the job status to Done before uploading the result, so we may have a pre-signed URL but no results uploaded yet.

What do you think?

Reply:
If we get HTTP errors after N retries why would we not raise the exception? Technically everything has gone wrong. The retries are the mechanism to try subdue bubbling exceptions but it would come off a tad unreliable if we return None to mask exceptions, no?